### ISSUE TITLE
FailFast in Ping

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#35421]
-2146232797 == COR_E_FAILFAST. Ping status is `Free` when it is asserted to be `InProgress`

Not obvoius there would be bad consequences in release, however.
```c#
        private void Finish()
        {
            lock (_lockObject)
            {
                Debug.Assert(_status == InProgress, $"Invalid status: {_status}"); <<<<<
```

```
C:\dotnetbuild\work\5ec212ee-4040-452f-af5d-b8d48facf2c2\Work\b0dfbbb7-ae0c-4187-b8eb-0890fd12c076\Exec>C:\dotnetbuild\work\5ec212ee-4040-452f-af5d-b8d48facf2c2\Payload\dotnet.exe xunit.console.dll System.Net.Ping.Functional.Tests.dll -xml testResults.xml -nologo -notrait category=nonnetcoreapptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=Outerloop 
  Discovering: System.Net.Ping.Functional.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Ping.Functional.Tests (found 36 of 46 test cases)
  Starting:    System.Net.Ping.Functional.Tests (parallel test collections = on, max threads = 2)
Assertion Failed
Invalid status: 0

   at System.Net.NetworkInformation.Ping.Finish() in D:\a\1\s\src\System.Net.Ping\src\System\Net\NetworkInformation\Ping.cs:line 126
   at System.Net.NetworkInformation.Ping.GetAddressAndSendAsync(String hostNameOrAddress, Int32 timeout, Byte[] buffer, PingOptions options) in D:\a\1\s\src\System.Net.Ping\src\System\Net\NetworkInformation\Ping.cs:line 395
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\AsyncMethodBuilder.cs:line 583
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\ExecutionContext.cs:line 173
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\AsyncMethodBuilder.cs:line 619
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext() in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\AsyncMethodBuilder.cs:line 598
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\TaskContinuation.cs:line 790
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 3354
   at System.Threading.Tasks.Task.FinishContinuations() in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 3328
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Future.cs:line 418
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\FutureFactory.cs:line 508
   at System.Threading.Tasks.TaskFactory`1.<>c__DisplayClass38_0`1.<FromAsyncImpl>b__0(IAsyncResult iar) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\FutureFactory.cs:line 904
   at System.Net.LazyAsyncResult.Complete(IntPtr userToken) in D:\a\1\s\src\Common\src\System\Net\LazyAsyncResult.cs:line 397
   at System.Net.ContextAwareResult.CompleteCallback() in D:\a\1\s\src\Common\src\System\Net\ContextAwareResult.cs:line 394
   at System.Net.ContextAwareResult.<>c.<Complete>b__17_0(Object s) in D:\a\1\s\src\Common\src\System\Net\ContextAwareResult.cs:line 388
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\ExecutionContext.cs:line 173
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\ExecutionContext.cs:line 133
   at System.Net.ContextAwareResult.Complete(IntPtr userToken) in D:\a\1\s\src\Common\src\System\Net\ContextAwareResult.cs:line 388
   at System.Net.LazyAsyncResult.ProtectedInvokeCallback(Object result, IntPtr userToken) in D:\a\1\s\src\Common\src\System\Net\LazyAsyncResult.cs:line 351
   at System.Net.LazyAsyncResult.InvokeCallback(Object result) in D:\a\1\s\src\Common\src\System\Net\LazyAsyncResult.cs:line 358
   at System.Net.NameResolutionPal.GetAddrInfoExState.<>c.<CompleteAsyncResult>b__5_0(Object s) in D:\a\1\s\src\System.Net.NameResolution\src\System\Net\NameResolutionPal.Windows.cs:line 455
   at System.Threading.Tasks.Task.InnerInvoke() in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 2484
   at System.Threading.Tasks.Task.<>c.<.cctor>b__277_0(Object obj) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 2463
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\ExecutionContext.cs:line 290
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 2424
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread threadPoolThread) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 2362
   at System.Threading.Tasks.Task.ExecuteFromThreadPool(Thread threadPoolThread) in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\Task.cs:line 2347
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in F:\vsagent\13\s\src\System.Private.CoreLib\shared\System\Threading\ThreadPool.cs:line 631
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback() in F:\vsagent\13\s\src\System.Private.CoreLib\src\System\Threading\ThreadPool.CoreCLR.cs:line 28
----- end 18:52:04.47 ----- exit code -2146232797 ----------------------------------------------------------

**Comment by danmosemsft** [comment]
This is not an outer loop test - it broke CI. 
@rmkerr is Ping yours?

**Comment by rmkerr** [comment]
I haven't worked on it before, but I'm not sure we have a designated owner. @wfurt, is this related to any of your recent ping work?

**Comment by davidsh** [comment]
Looks like a dupe of #35351

**Comment by danmosemsft** [comment]
Is not obvious to me it's a dupe?

**Comment by davidsh** [comment]
>Is not obvious to me it's a dupe?

I think it is a dupe because "Invalid status: 0" is basically the same as the other error code from the other issue ("The operation completed successfully.").  Usually the "success" error codes have a value of zero.

@wfurt is working on Ping API in general and he will be the one to confirm root cause.

**Comment by wfurt** [comment]
The code did not changed very recently. It seems like this is failing often now. This may be old bug exposed by some environmental changes.

**Comment by wfurt** [comment]
fixed by #33621 and dotnet/coreclr#22871
